### PR TITLE
dev/core#321: Prevent duplicate entries in civicrm_entity_file

### PR DESCRIPTION
**Pull request by monishdeb**
Overview
----------------------------------------
Steps to replicate:

Enable CiviCase
Create a new Case.
Edit any Case Activity >> add an attachment >> Save
Reopen the Edit Activity form and simply save.

DB result: There will be duplicate entries in civicrm_entity_file

Before
----------------------------------------
Duplicate entries in  civicrm_entity_file

After
----------------------------------------
No duplicate entries in  civicrm_entity_file

Technical Details
----------------------------------------
This PR introduces:
1.  Unique Key to civicrm_entity_file table and handled in install and 5.5alpha1 upgrade
2. Added new function to create unique index ```CRM_Core_BAO_SchemaHandler::createUniqueIndex(...)```

### CODE REVIEW

**Comment by monishdeb** [comment]
ping @lcdservices @colemanw

**Comment by lcdservices** [comment]
@monishdeb tested and confirmed. entity_file records are no longer being duplicated.

**Comment by eileenmcnaughton** [comment]
@monishdeb I'm happy to merge the change to CRM/Case/Form/Activity.php as it has been teted & the schema changes as they affect new installs. If you split out that part I'll merge

 But I think sites should be using the api 

System.updateindexes to add any missing indexes - can you see if that job will add your index & if not add tests / steps for that.

**Comment by eileenmcnaughton** [comment]
@monishdeb per comments - if you remove the indexes upgrade portion here I will merge

**Comment by monishdeb** [comment]
@eileenmcnaughton sorry for delay. Removed the indexes upgrade code. Please have a look

File: CRM/Upgrade/Incremental/php/FiveFive.php [hunk 1]
```diff
@@ -46,40 +46,6 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
...
-   * dev/core#321 - Prevent duplicate entries in civicrm_entity_file
-   *
-   * @param \CRM_Queue_TaskContext $ctx
-   *
-   * @return bool
-   */
-  public static function entityFileSchemaUpdate(CRM_Queue_TaskContext $ctx) {
-    if (CRM_Core_BAO_SchemaHandler::checkIfIndexExists('civicrm_entity_file', 'index_entity_file_id')) {
-      // Delete any stray duplicate rows and add unique index to prevent new dupes and enable INSERT/UPDATE combo query
-      CRM_Core_DAO::executeQuery('
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FiveFive.php]: @monishdeb so I don't want to add the indices but quite possibly this cleanup should still run?  I assume the data here is still invalid?

Also, we should squash before merging

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Just revisiting this - ooks like it was pretty much agreed it just got stalled

**Pull request merged**